src/transformers/models/chatglm/modeling_chatglm.py

```diff
@@ -0,0 +1,1289 @@
+        self.register_buffer("sin_cached", emb.sin().to(dtype), persistent=False)
+
+    def forward(self, x, seq_len=None):
+        # x: [bs, num_attention_heads, seq_len, head_size]
+        if seq_len > self.max_seq_len_cached:
+            self._set_cos_sin_cache(seq_len=seq_len, device=x.device, dtype=x.dtype)
+
+        return (
+            self.cos_cached[:seq_len].to(dtype=x.dtype),
+            self.sin_cached[:seq_len].to(dtype=x.dtype),
+        )
+
+
+class ChatGlmAttention(nn.Module):
+    def __init__(self, config, layer_idx: Optional[int] = None):
+        super().__init__()
+
+        self.config = config
+        self.apply_query_key_layer_scaling = config.apply_query_key_layer_scaling
+        self.hidden_size = config.hidden_size
+        self.num_heads = config.num_attention_heads
+        self.head_dim = self.hidden_size // self.num_heads
+        self.split_size = self.hidden_size
+        self.multi_query_attention = config.multi_query_attention
+        self.max_position_embeddings = config.max_position_embeddings
+
+        self.projection_size = config.kv_channels * config.num_attention_heads
+        self.hidden_size_per_attention_head = self.projection_size // config.num_attention_heads
+
+        self.layer_idx = layer_idx
+        if layer_idx is None:
+            logger.warning_once(
+                f"Instantiating {self.__class__.__name__} without passing `layer_idx` is not recommended and will "
+                "to errors during the forward call, if caching is used. Please make sure to provide a `layer_idx` "
+                "when creating this class."
+            )
+
+        if self.head_dim * self.num_heads != self.hidden_size:
+            raise ValueError(
+                f"`hidden_size` must be divisible by num_heads (got `hidden_size`: {self.hidden_size} and `num_heads`:"
+                f" {self.num_heads})."
+            )
+
+        # Layer-wise attention scaling it is important to keep this for numerical stability
+        # to get the same results as the original `trust_remote_code` model.
+        if self.apply_query_key_layer_scaling:
+            self.coeff = layer_idx + 1
+        else:
+            self.coeff = 1
+
+        # in the `trust_remote_code` model the inverse norm factor is multiplied
+        # by `self.coeff` and `self.coeff` is used back when computing the final
+        # attention scores
+        self.inv_norm_factor = 1.0 / (math.sqrt(self.head_dim) * self.coeff)
+        self.partial_rotary_factor = config.partial_rotary_factor
+
+        self.rope_theta = config.rope_theta
+
+        if not self.multi_query_attention:
+            qkv_size = 3 * self.hidden_size
+            self.num_key_value_heads = 1
+        else:
+            self.num_key_value_heads = config.multi_query_group_num
+            qkv_size = self.projection_size + 2 * self.hidden_size_per_attention_head * self.num_key_value_heads
```

Comment:
we should just have a general implementation! supporting GQA means you can have MQA, MHA and GQA

Reply:
agreed!

Reply:
I might have time to work on the addition next week, with a new cool feature that will help ease the integration! 🤗